myia/abstract/map.py

```diff
@@ -366,6 +369,12 @@ def abstract_map2(  # noqa: F811
     if len(x.elements) != len(y.elements):
         raise MapError(x, y, reason="Structures have different lengths")
 
+    if x.tracks.interface is dict:
+        # Just check keys, tracks will be checked below.
+        for kx, ky in zip(x.elements[::2], y.elements[::2]):
+            if kx is not ky:
+                raise MapError(kx, ky, "Different dict keys")
+
```

Comment:
The keys should be in the interface, though (e.g. a class `DictWithKeys("x", "y")`), and if the keys are different, the interfaces would be different, so there would be no need to special case anything.